flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -511,9 +511,8 @@ public void deploy() throws JobException {
         }
 
         // make sure exactly one deployment call happens from the correct state
-        // note: the transition from CREATED to DEPLOYING is for testing purposes only
         ExecutionState previous = this.state;
-        if (previous == SCHEDULED || previous == CREATED) {
```

Comment:
I'm wondering whether it would make sense to define valid transitions in the `ExecutionState` and then have something along the lines of `ExecutionState#isAllowedToTransitionTo(ExecutionState)`. That would introduce a single place with the state machine definition that could guard the contract 🤔.

Reply:
I had similar thought. It's not possible in the ExecutionState (because of circular dependencies between RUNNING and RESTARTING), but we could certainly introduce a separate ExecutionStateTransition enum that captures all of that.

Reply:
Or rather it could be written in a nicer way as a separate enum.

Reply:
Or even a plain EnumMap